ls1intum/Artemis
# `Development`: Validate example solution publication date

### Checklist
#### General


- [x] This was confirmed by another developer on a test server.
#### Server
#### Client

### Motivation and Context


Example Solution Publication Date did not have a min date constraint, this could lead to potential mistakes like publishing a solution before the exercise is over or even before it begins.
### Description

Example Solution Publication Date of an exercise should be after Release Date and Due Date if they are set. It is allowed to be before the Due Dateif the exercise is optional, i.e. not included in grading. It can also be empty. This validation is made both at client and the server.

Also, at client side the user is now warned if they set the Example Solution Publication Date before the Due Date for an optional exercise in order to prevent early publications of the solution accidentally.

Thanks to @Hialus we noticed that date validation is not performed at all for programming exercises on updates, it is only done during inserts. This is also fixed in this PR.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise
- 1 Text Exercise
- 1 Modeling Exercise
- 1 File Upload Exercise

1. Log in to Artemis
2. Navigate to Course Administration
3. For each exercise type above, try the following:
4. Save the exercise with an empty Example Solution Publication Date when release date or due date is set and see it succeed.
5. Save the exercise with an Example Solution Publication Date after the due date and see it succeed.
6. Save the exercise with empty Example Solution Publication Date, release date, due date and see it succeed.
7. Unset the due date and save the exercise with an Example Solution Publication Date after the release date and see it succeed.
8. Set `Should this exercise be included in the course score calculation?` to either `Yes`  or `Bonus`, then try to save the exercise with an Example Solution Publication Date before due date and see it fail with a date validation error.
9. Set `Should this exercise be included in the course score calculation?` to `No`, then try to save the exercise with an Example Solution Publication Date before due date (after release date if set) and see it succeed with a yellow warning alert shown after save button is pressed.
10. Unset the due date and try to save the exercise with an Example Solution Publication Date before release date and see it fail with a date validation error.
### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| BaseExercise.java | 89% | 98% |
| exercise.model.ts | 97.14% | 98.75% |
| file-upload-exercise-update.component.ts | 77.5% | 85% |
| modeling-exercise-update.component.ts | 80.95% | 91.26% |
| programming-exercise-lifecycle.component.ts | 77.77% | 94.59% |
| exercise.service.ts | 82.2% | (did not add tests for just http calls ) 73.15% |
| exercise.utils.ts | 86.45% | 90.9% |
| text-exercise-update.component.ts | 84.41% | 91.39% |
| Exercise.java | 82% | 93% |
| ProgrammingExerciseService.java | 81% | 92% |
| ProgrammingExerciseResource.java | 81% | 90% |

### Screenshots



![image](https://user-images.githubusercontent.com/18427209/169002290-3e0794c3-281f-4306-845f-b08466130015.png)

![image](https://user-images.githubusercontent.com/18427209/168967299-0f53440b-ae40-48cd-8e92-70112c7f0a9a.png)
